File: implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/user/UserInsertSetStepSetterTest.java
package ch.difty.scipamato.core.persistence.user;

import static ch.difty.scipamato.core.db.tables.ScipamatoUser.SCIPAMATO_USER;
import static ch.difty.scipamato.core.persistence.user.UserRecordMapperTest.*;
import static org.mockito.Mockito.*;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;

import ch.difty.scipamato.core.db.tables.records.ScipamatoUserRecord;
import ch.difty.scipamato.core.entity.User;
import ch.difty.scipamato.core.persistence.InsertSetStepSetter;
import ch.difty.scipamato.core.persistence.InsertSetStepSetterTest;

@SuppressWarnings("ResultOfMethodCallIgnored")
class UserInsertSetStepSetterTest extends InsertSetStepSetterTest<ScipamatoUserRecord, User> {

    private final InsertSetStepSetter<ScipamatoUserRecord, User> setter = new UserInsertSetStepSetter();

    @Mock
    private User entityMock;

    @Mock
    private ScipamatoUserRecord recordMock;

    @Override
    protected InsertSetStepSetter<ScipamatoUserRecord, User> getSetter() {
        return setter;
    }

    @Override
    protected User getEntity() {
        return entityMock;
    }

    @Override
    protected void specificTearDown() {
        verifyNoMoreInteractions(entityMock, recordMock);
    }

    @Override
    protected void entityFixture() {
        UserRecordMapperTest.entityFixtureWithoutIdFields(entityMock);
    }

    @Override
    protected void stepSetFixtureExceptAudit() {
        doReturn(getMoreStep())
            .when(getStep())
            .set(SCIPAMATO_USER.USER_NAME, USER_NAME);
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.FIRST_NAME, FIRST_NAME);
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.LAST_NAME, LAST_NAME);
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.EMAIL, EMAIL);
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.PASSWORD, PASSWORD);
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.ENABLED, ENABLED);
    }

    @Override
    protected void setStepFixtureAudit() {
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.CREATED_BY, UserRecordMapperTest.CREATED_BY);
        doReturn(getMoreStep())
            .when(getMoreStep())
            .set(SCIPAMATO_USER.LAST_MODIFIED_BY, UserRecordMapperTest.LAST_MOD_BY);
    }

    @Override
    protected void verifyCallToFieldsExceptKeyAndAudit() {
        verify(entityMock).getUserName();
        verify(entityMock).getFirstName();
        verify(entityMock).getLastName();
        verify(entityMock).getEmail();
        verify(entityMock).getPassword();
        verify(entityMock).isEnabled();
    }

    @Override
    protected void verifySettingFieldsExceptKeyAndAudit() {
        verify(getStep()).set(SCIPAMATO_USER.USER_NAME, USER_NAME);

        verify(getMoreStep()).set(SCIPAMATO_USER.FIRST_NAME, FIRST_NAME);
        verify(getMoreStep()).set(SCIPAMATO_USER.LAST_NAME, LAST_NAME);
        verify(getMoreStep()).set(SCIPAMATO_USER.EMAIL, EMAIL);
        verify(getMoreStep()).set(SCIPAMATO_USER.PASSWORD, PASSWORD);
        verify(getMoreStep()).set(SCIPAMATO_USER.ENABLED, ENABLED);
    }

    @Override
    protected void verifySettingAuditFields() {
        verify(getMoreStep()).set(SCIPAMATO_USER.CREATED_BY, UserRecordMapperTest.CREATED_BY);
        verify(getMoreStep()).set(SCIPAMATO_USER.LAST_MODIFIED_BY, UserRecordMapperTest.LAST_MOD_BY);
    }

    @Test
    void consideringSettingKeyOf_withNullId_doesNotSetId() {
        when(getEntity().getId()).thenReturn(null);
        getSetter().considerSettingKeyOf(getMoreStep(), getEntity());
        verify(getEntity()).getId();
    }

    @Test
    void consideringSettingKeyOf_withNonNullId_doesSetId() {
        when(getEntity().getId()).thenReturn(ID);

        getSetter().considerSettingKeyOf(getMoreStep(), getEntity());

        verify(getEntity()).getId();
        verify(getMoreStep()).set(SCIPAMATO_USER.ID, ID);
    }

    @Test
    void resettingIdToEntity_withNullRecord_doesNothing() {
        getSetter().resetIdToEntity(entityMock, null);
        verify(entityMock, never()).setId(anyInt());
    }

    @Test
    void resettingIdToEntity_withNonNullRecord_setsId() {
        when(recordMock.getId()).thenReturn(3);
        getSetter().resetIdToEntity(entityMock, recordMock);
        verify(recordMock).getId();
        verify(entityMock).setId(anyInt());
    }

}

File: implementation/scipamato/core/persistence-jooq/src/test/kotlin/ch/difty/scipamato/core/persistence/user/UserInsertSetStepSetterTest.kt
package ch.difty.scipamato.core.persistence.user

import ch.difty.scipamato.core.db.tables.ScipamatoUser.SCIPAMATO_USER
import ch.difty.scipamato.core.db.tables.records.ScipamatoUserRecord
import ch.difty.scipamato.core.entity.User
import ch.difty.scipamato.core.persistence.InsertSetStepSetter
import ch.difty.scipamato.core.persistence.InsertSetStepSetterTest
import com.nhaarman.mockitokotlin2.*
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyInt

internal class UserInsertSetStepSetterTest : InsertSetStepSetterTest<ScipamatoUserRecord, User>() {

    override val setter: InsertSetStepSetter<ScipamatoUserRecord, User> = UserInsertSetStepSetter()

    override val entity = mock<User>()
    private val recordMock = mock<ScipamatoUserRecord>()

    override fun specificTearDown() {
        verifyNoMoreInteractions(entity, recordMock)
    }

    override fun entityFixture() {
        UserRecordMapperTest.entityFixtureWithoutIdFields(entity)
    }

    override fun stepSetFixtureExceptAudit() {
        doReturn(moreStep).whenever(step).set(SCIPAMATO_USER.USER_NAME, UserRecordMapperTest.USER_NAME)
        doReturn(moreStep).whenever(moreStep).set(SCIPAMATO_USER.FIRST_NAME, UserRecordMapperTest.FIRST_NAME)
        doReturn(moreStep).whenever(moreStep).set(SCIPAMATO_USER.LAST_NAME, UserRecordMapperTest.LAST_NAME)
        doReturn(moreStep).whenever(moreStep).set(SCIPAMATO_USER.EMAIL, UserRecordMapperTest.EMAIL)
        doReturn(moreStep).whenever(moreStep).set(SCIPAMATO_USER.PASSWORD, UserRecordMapperTest.PASSWORD)
        doReturn(moreStep).whenever(moreStep).set(SCIPAMATO_USER.ENABLED, UserRecordMapperTest.ENABLED)
    }

    override fun setStepFixtureAudit() {
        doReturn(moreStep).whenever(moreStep)
            .set(SCIPAMATO_USER.CREATED_BY, ch.difty.scipamato.core.persistence.RecordMapperTest.CREATED_BY)
        doReturn(moreStep).whenever(moreStep)
            .set(SCIPAMATO_USER.LAST_MODIFIED_BY, ch.difty.scipamato.core.persistence.RecordMapperTest.LAST_MOD_BY)
    }

    override fun verifyCallToFieldsExceptKeyAndAudit() {
        verify(entity).userName
        verify(entity).firstName
        verify(entity).lastName
        verify(entity).email
        verify(entity).password
        verify(entity).isEnabled
    }

    override fun verifySettingFieldsExceptKeyAndAudit() {
        verify(step).set(SCIPAMATO_USER.USER_NAME, UserRecordMapperTest.USER_NAME)

        verify(moreStep).set(SCIPAMATO_USER.FIRST_NAME, UserRecordMapperTest.FIRST_NAME)
        verify(moreStep).set(SCIPAMATO_USER.LAST_NAME, UserRecordMapperTest.LAST_NAME)
        verify(moreStep).set(SCIPAMATO_USER.EMAIL, UserRecordMapperTest.EMAIL)
        verify(moreStep).set(SCIPAMATO_USER.PASSWORD, UserRecordMapperTest.PASSWORD)
        verify(moreStep).set(SCIPAMATO_USER.ENABLED, UserRecordMapperTest.ENABLED)
    }

    override fun verifySettingAuditFields() {
        verify(moreStep).set(SCIPAMATO_USER.CREATED_BY, ch.difty.scipamato.core.persistence.RecordMapperTest.CREATED_BY)
        verify(moreStep)
            .set(SCIPAMATO_USER.LAST_MODIFIED_BY, ch.difty.scipamato.core.persistence.RecordMapperTest.LAST_MOD_BY)
    }

    @Test
    fun consideringSettingKeyOf_withNullId_doesNotSetId() {
        whenever(entity.id).thenReturn(null)
        setter.considerSettingKeyOf(moreStep, entity)
        verify<User>(entity).id
    }

    @Test
    fun consideringSettingKeyOf_withNonNullId_doesSetId() {
        whenever(entity.id).thenReturn(UserRecordMapperTest.ID)
        setter.considerSettingKeyOf(moreStep, entity)
        verify<User>(entity).id
        verify(moreStep).set(SCIPAMATO_USER.ID, UserRecordMapperTest.ID)
    }

    @Test
    fun resettingIdToEntity_withNullRecord_doesNothing() {
        setter.resetIdToEntity(entity, null)
        verify(entity, never()).id = anyInt()
    }

    @Test
    fun resettingIdToEntity_withNonNullRecord_setsId() {
        whenever(recordMock.id).thenReturn(3)
        setter.resetIdToEntity(entity, recordMock)
        verify(recordMock).id
        verify<User>(entity).id = anyInt()
    }
}
